atuin-client/src/history.rs

```diff
@@ -5,38 +5,62 @@ use serde::{Deserialize, Serialize};
 
 use atuin_common::utils::uuid_v7;
 
-// Any new fields MUST be Optional<>!
+mod builder;
+
+/// A marker type used to seal the `History` struct, preventing it from being constructed directly.
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub(self) struct HistorySeal;
+
+/// Client-side history entry.
+///
+/// Client stores data unencrypted, and only encrypts it before sending to the server.
+///
+/// To create a new history entry, use one of the builders:
+/// - [`History::import()`] to import an entry from the shell history file
+/// - [`History::capture()`] to capture an entry via hook
+/// - [`History::from_db()`] to create an instance from the database entry
+///
+/// ### Caution
+/// Any new fields MUST be `Optional<T>` and marked with `#[serde(default)]` to ensure backwards
+/// compatibility with older clients.
 #[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, sqlx::FromRow)]
 pub struct History {
+    /// A client-generated ID, used to identify the entry when syncing.
+    ///
+    /// Stored as `client_id` in the database.
     pub id: String,
+    /// When the command was run.
     pub timestamp: chrono::DateTime<Utc>,
+    /// How long the command took to run.
     pub duration: i64,
+    /// The exit code of the command.
     pub exit: i64,
+    /// The command that was run.
     pub command: String,
+    /// The current working directory when the command was run.
     pub cwd: String,
+    /// The session ID, associated with a terminal session.
     pub session: String,
+    /// The hostname of the machine the command was run on.
     pub hostname: String,
+
+    /// Timestamp, which is set when the entry is deleted, allowing a soft delete.
+    #[serde(default)]
     pub deleted_at: Option<chrono::DateTime<Utc>>,
-}
 
-// Forgive me, for I have sinned
-// I need to replace rmp with something that is more backwards-compatible.
-// Protobuf, or maybe just json
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, sqlx::FromRow)]
-pub struct HistoryWithoutDelete {
```

Comment:
We don't need such a struct with right serde attributes. Down below are the tests, confirming it.

cc @conradludgate regarding removal of HistoryWithoutDelete

Reply:
Hey! This is _intentionally_ horrible right now. We don't want to make it nice and easy to add new fields!

The reasoning here is that I'd like to change our serialization format to one that better supports versioning and migrations. By leaving it awkward, it'll force our hand in the future when we wish to add new fields. Making it "good enough" means that this will not happen.

Reply:
Ok, so in the initial PR we do want to add a new field :)

Would that be ok if I also change serialization format here, then?

Reply:
And in this PR I introduced failing test, so that we won't forget that migration is indeed broken in one direction: https://github.com/ellie/atuin/pull/933#discussion_r1183559996

Reply:
I changed the mechanism to make how we encode histories for encryption to be explicit. This is the best of both worlds. `serde` is not designed for non self describing documents so it should not be used moving forward.

#1038 is using rmp-serde now too, so I will remove it from there in a separate PR.